alpaca_trade_api/rest_async.py

```diff
@@ -0,0 +1,183 @@
+            "end":   end,
+            "limit": limit,
+        }
+        df = await self._iterate_requests(symbol, payload, limit, _type,
+                                          TradesV2)
+
+        return df
+
+    async def get_quotes_async(self, symbol, start, end, timeframe,
+                               limit=1000):
+        _type = "quotes"
+
+        payload = {
+            "start": start,
+            "end":   end,
+            "limit": limit,
+        }
+        df = await self._iterate_requests(symbol, payload, limit, _type,
+                                          QuotesV2)
+
+        return df
+
+    async def get_latest_trade_async(self, symbol: str) -> TradeV2:
+        """
+        Get the latest trade for the given symbol
+        """
+        _type = "trades"
+        url = self._get_latest_url(_type, symbol)
+        opts = self._get_opts()
+
+        async with aiohttp.ClientSession() as session:
+            async with session.get(url, **opts) as response:
+                response = await response.json()
+                if response.get("quote"):
+                    result = TradeV2(response["trade"])
+                    return result
+
+    async def get_latest_quote_async(self, symbol: str) -> QuoteV2:
+        """
+        Get the latest trade for the given symbol
+        """
+        _type = "quotes"
+        url = self._get_latest_url(_type, symbol)
+        opts = self._get_opts()
+
+        async with aiohttp.ClientSession() as session:
+            async with session.get(url, **opts) as response:
+                response = await response.json()
+                if response.get("quote"):
+                    result = QuoteV2(response["quote"])
+                    return result
+
+    def _get_opts(self, payload=None):
+        headers = {}
+        headers['APCA-API-KEY-ID'] = self._key_id
+        headers['APCA-API-SECRET-KEY'] = self._secret_key
+        opts = {
+            'headers':         headers,
+            # Since we allow users to set endpoint URL via env var,
+            # human error to put non-SSL endpoint could exploit
+            # uncanny issues in non-GET request redirecting http->https.
+            # It's better to fail early if the URL isn't right.
+            'allow_redirects': False,
+        }
+        opts['params'] = payload
+
+        return opts
+
+    async def _request(self, symbol, url, payload):
+        opts = self._get_opts(payload)
+        async with aiohttp.ClientSession() as session:
+            try:
+                while 1:
+                    async with session.get(url, **opts) as response:
+
+                        response = await response.json()
+                        page_token = response.get('next_page_token')
+                        payload["page_token"] = page_token
+                        yield response
+
+                        if not page_token:
+                            break
+            except Exception as e:
```

Comment:
I really don't like handling all exceptions like this.

Reply:
the thing is that this method is used for bulk data retrieve. if I want to get data for 2000 symbols, if one fails I don't want to fail everything. I want to provide best effort.
for instance, if I remove this in my example when waiting for tasks to complete (in this code: `results = await asyncio.gather(*tasks)`

I get this exception for one symbol: 

![image](https://user-images.githubusercontent.com/78740792/132027782-13da5ea6-fee8-4697-81d4-7774c636cdf4.png)

that fails everything.

do you have a better suggestion to allow getting the "best effort" outcome?

Reply:
I would let the caller decide how he wants to handle the exception. In this particular case I would
- add `return_exceptions=True` to the `asyncio.gather` call(s)
- when iterating through the results, check if it's an exception